### PR TITLE
Docs improvements around hparams

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #8525

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]

File: docs/source/common/hyperparameters.rst [hunk 1]
```diff
@@ -152,22 +152,25 @@ improve readability and reproducibility.
...
 
     .. code-block:: python
 
-        # using a argparse.Namespace
         class LitMNIST(LightningModule):
 
-            def __init__(self, conf, *args, **kwargs):
+            def __init__(self, conf: Union[Dict, Namespace, DictConfig], *args, **kwargs):
                 super().__init__()
                 self.save_hyperparameters(conf)
```
Comment (awaelchli) [inline docs/source/common/hyperparameters.rst:163]: I just noticed, I think these docs are incomplete. While this works for purely saving the parameters, it will not be possible to restore the model because it's ambiguous how to pass arguments to the constructor. 

One should either unroll the config in the signature, i.e., `def __init__(**conf)` or save it like so: `self.save_hyperparameters("conf")`

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source/common/hyperparameters.rst [hunk 2]
```diff
@@ -153,28 +153,34 @@ improve readability and reproducibility.
...
 
-3.  You can also save full objects such as `dict` or `Namespace` to the checkpoint.
+3.  You can also convert full objects such as ``dict`` or ``Namespace`` to ``hparams`` so they get saved to the
+    checkpoint.
 
     .. code-block:: python
 
-        # using a argparse.Namespace
         class LitMNIST(LightningModule):
             def __init__(self, conf, *args, **kwargs):
```
Comment (carmocca) [inline docs/source/common/hyperparameters.rst:162]:
```suggestion
            def __init__(self, conf: Optional[Union[Dict, Namespace, DictConfig]] = None, **kwargs):
```

Comment (Borda) [reply docs/source/common/hyperparameters.rst]: `def __init__(self, conf: Optional[Union[Dict, Namespace, DictConfig]] = None, **kwargs: Dict[str, Any]):`

File: pytorch_lightning/utilities/parsing.py [hunk 3]
```diff
@@ -253,29 +253,24 @@ def save_hyperparameters(
...
-        try:
-            return self[key]
-        except KeyError as exp:
-            raise AttributeError(f'Missing attribute "{key}"') from exp
-
-    def __setattr__(self, key: str, val: Any) -> None:
-        self[key] = val
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self.__dict__ = self
```
Comment (tchaton) [inline pytorch_lightning/utilities/parsing.py:273]: quite ingenious !

File: docs/source/common/hyperparameters.rst [hunk 4]
```diff
@@ -116,8 +116,8 @@ at which point it is very useful to know how that model was trained (i.e.: what
 Lightning has a few ways of saving that information for you in checkpoints and yaml files. The goal here is to
 improve readability and reproducibility.
 
-1.  The first way is to ask lightning to save the values of anything in the __init__ for you to the checkpoint. This also
-    makes those values available via `self.hparams`.
+1.  The first way is to ask lightning to save the values of anything in the ``__init__`` for you to the checkpoint.
```
Comment (tchaton) [inline docs/source/common/hyperparameters.rst:119]:
```suggestion
1. Using :meth:`~pytorch_lightning.core.lightning.LightningModule. save_hyperparameters` within your :class:`~pytorch_lightning.core.lightning.LightningModule` ``__init__`` function will enable Lightning to save all th e provided arguments which will be stored within the model checkpoint. This simplify your model re-instantiation in production settings.
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source/common/hyperparameters.rst [hunk 5]
```diff
@@ -116,8 +116,11 @@ at which point it is very useful to know how that model was trained (i.e.: what
 Lightning has a few ways of saving that information for you in checkpoints and yaml files. The goal here is to
 improve readability and reproducibility.
 
-1.  The first way is to ask lightning to save the values of anything in the __init__ for you to the checkpoint. This also
-    makes those values available via `self.hparams`.
+1. Using :meth:`~pytorch_lightning.core.lightning.LightningModule. save_hyperparameters` within your
```
Comment (carmocca) [inline docs/source/common/hyperparameters.rst:119]:
```suggestion
1.  Using :meth:`~pytorch_lightning.core.lightning.LightningModule. save_hyperparameters` within your
```

**Comment by carmocca** [comment]
Removed the `AttributeDict` changes as it seems like it was leaving leftover CUDA memory which wasn't getting cleared.

In the future, we should try and replace its usage with `dataclasses`
